modules/renter/contractor/renew.go

```diff
@@ -0,0 +1,127 @@
+	}
+	ourAddress := c.cachedAddress
+	c.mu.Unlock()
+
+	renterCost := host.Price.Mul(types.NewCurrency64(contract.LastRevision.NewFileSize)).Mul(types.NewCurrency64(uint64(newEndHeight - height)))
+	renterCost = renterCost.MulFloat(1.05) // extra buffer to guarantee we won't run out of money during revision
+	payout := renterCost                   // no collateral
+
+	// create file contract
+	fc := types.FileContract{
+		FileSize:       contract.LastRevision.NewFileSize,
+		FileMerkleRoot: contract.LastRevision.NewFileMerkleRoot,
+		WindowStart:    newEndHeight,
+		WindowEnd:      newEndHeight + host.WindowSize,
+		Payout:         payout,
+		UnlockHash:     types.UnlockHash{}, // to be filled in by negotiateContract
+		RevisionNumber: 0,
+		ValidProofOutputs: []types.SiacoinOutput{
+			// nothing returned to us; everything goes to the host
+			{Value: types.ZeroCurrency, UnlockHash: ourAddress},
+			{Value: types.PostTax(height, renterCost), UnlockHash: host.UnlockHash},
+		},
+		MissedProofOutputs: []types.SiacoinOutput{
+			// nothing returned to us; everything goes to the void
+			{Value: types.ZeroCurrency, UnlockHash: ourAddress},
+			{Value: types.PostTax(height, renterCost), UnlockHash: types.UnlockHash{}},
+		},
+	}
+
+	// create transaction builder
+	txnBuilder := c.wallet.StartTransaction()
+
+	// initiate connection
+	conn, err := c.dialer.DialTimeout(contract.IP, 15*time.Second)
+	if err != nil {
+		return types.FileContractID{}, err
+	}
+	defer conn.Close()
+	if err := encoding.WriteObject(conn, modules.RPCRenew); err != nil {
+		return types.FileContractID{}, errors.New("couldn't initiate RPC: " + err.Error())
+	}
+	if err := encoding.WriteObject(conn, fcid); err != nil {
+		return types.FileContractID{}, errors.New("couldn't send contract ID: " + err.Error())
+	}
+
+	// execute negotiation protocol
+	newContract, err := negotiateContract(conn, contract.IP, fc, txnBuilder, c.tpool)
+	if err != nil {
+		txnBuilder.Drop() // return unused outputs to wallet
+		return types.FileContractID{}, err
+	}
+
+	// update host contract
+	c.mu.Lock()
+	c.contracts[newContract.ID] = newContract
+	c.cachedAddress = types.UnlockHash{} // clear cachedAddress
+	err = c.save()
+	c.mu.Unlock()
+	if err != nil {
+		c.log.Println("WARN: failed to save the contractor:", err)
+	}
+
+	return newContract.ID, nil
+}
+
+// threadedRenewContracts renews the Contractor's contracts according to the
+// specified allowance and at the specified height.
+func (c *Contractor) threadedRenewContracts(allowance modules.Allowance, newHeight types.BlockHeight) {
+	var wg sync.WaitGroup
+	c.mu.RLock()
+	wg.Add(len(c.contracts))
+	for _, contract := range c.contracts {
+		if contract.FileContract.WindowStart < newHeight {
+			go func() {
+				defer wg.Done()
+				_, err := c.managedRenew(contract.ID, newHeight)
+				if err != nil {
+					c.log.Println("WARN: failed to renew contract", contract.ID, ":", err)
+				}
+			}()
+		}
+	}
+	c.mu.RUnlock()
```

Comment:
so, if I'm understanding correctly here, all the `managedRenew` calls are going to block before this unlock is called, and then afterwards Wait() is called to let all of the renewals take place. That should be in a comment somewhere.

I'm also not sure that it makes sense to lump all of the renewals into a single call? It seems like you could end up with multiple calls to `threadedRenewContracts` open at the same time, which could cause problems.